File: JavaScript/dijkstra_algorithm.js
Array.prototype.remove = Array.prototype.remove || function(element) {
    var index = this.indexOf(element);

    if (index >= 0) {
        this.splice(index, 1);
    };
};

Array.prototype.contains = Array.prototype.contains || function(element) {
    return this.indexOf(element) >= 0;
};

var graphs = {};

(function(host) {
    function Graph(nodesNumber, edges) {
        this.nodesNumber = nodesNumber;
        this.initEdges(edges);
    };

    Graph.prototype.initEdges = function(edges) {
        var oThis = this,
            i = 0;

        this.edges = [];
        for (; i < this.nodesNumber; i++) {
            this.edges[i] = [];
        };        
        if (edges) {
            edges.forEach(function (edge) {
                oThis.edge(edge[0], edge[1], edge[2]);
            });
        };
    };

    Graph.prototype.edge = function(from, to, weight) {
        this.edges[from - 1][to - 1] = weight;
        return this;
    };
    
    Graph.prototype._constructShortestPath = function(distances, previous, unvisited, to) {
        var vertex = to,
        path = [];

        while (undefined != vertex) {
            path.unshift(vertex + 1);
            vertex = previous[vertex];
        };
            
        return {
            path: path,
            length: distances[to]
        };
    };

    Graph.prototype._getUnvisitedVertexWithShortestPath = function(distances, previous, unvisited) {
        var minimumDistance = Number.MAX_VALUE,
            vertex = null;
            
        unvisited.forEach(function (unvisitedVertex) {
            if (distances[unvisitedVertex] < minimumDistance) {
                vertex = unvisitedVertex;
                minimumDistance = distances[vertex];
            };
        });
        return vertex;
    };
    
    Graph.prototype._updateDistancesForCurrent = function(distances, previous, unvisited, current) {    
        for (var i = 0; i < this.edges[current].length; i++) {
            var currentEdge = this.edges[current][i];
            
            if ((undefined != currentEdge) && unvisited.contains(i)) {
                if (distances[current] + currentEdge < distances[i]) {
                    distances[i] = distances[current] + currentEdge;
                    previous[i] = current;
                };
            };            
        };
    };

    //Dijkstra algorithm http://en.wikipedia.org/wiki/Dijkstra's_algorithm
    Graph.prototype.getShortestPath = function(from, to) {
        var unvisited = [],
            current = null,
            distances = [],
            previous = [];

        from = from - 1;        
        to = to - 1;
        //Initialization
        for (var i = 0; i < this.nodesNumber; i++) {
            unvisited.push(i);
            //Infinity
            distances.push(Number.MAX_VALUE);
        };
        distances[from] = 0;
        
        while (true) {
            if (!unvisited.contains(to)) {
                return this._constructShortestPath(distances, previous, unvisited, to);
            };
            current = this._getUnvisitedVertexWithShortestPath(distances, previous, unvisited);
        
            //No path exists
            if ((null == current) || (Number.MAX_VALUE == distances[current])) {
                return {
                    path: [],
                    length: Number.MAX_VALUE
                };
            };
            this._updateDistancesForCurrent(distances, previous, unvisited, current);            
            unvisited.remove(current);
        };
    };
    
    Graph.prototype.toString = function() {
        return this.edges.map(function (edgesFromVertex) {
            return edgesFromVertex.join(",")
        }).join("\n");        
    };
    host.Graph = Graph;
})(graphs);

var graph = new graphs.Graph(8, [
    [1, 2, 5], [1, 3, 1], [1, 4, 3],
    [2, 3, 2], [2, 5, 2],
    [3, 4, 1], [3, 5, 8],
    [4, 6, 2],
    [5, 7, 1],
    [6, 5, 1]
]);

var shortestPath = graph.getShortestPath(1, 7);

console.log("path = ", shortestPath.path.join(","));
console.log("length = ", shortestPath.length);

//No shortest path to the vertex '8'
console.log(graph.getShortestPath(1, 8));

File: Dart/dijkstra_algorithm.dart
library Graphs;

class Graph {
  num nodesNumber;
  List<List<num>> edges;
  
  Graph(num nodesNumber, List<List<num>> edges) {
      this.nodesNumber = nodesNumber;
      initEdges(edges);
  }

  void initEdges(List<List<num>> edges) {
      this.edges = new List<List<num>>();
      for (int i = 0; i < nodesNumber; i++) {
          List<num> row = new List<num>();

          for (int j = 0; j < nodesNumber; j++) {            
              row.add(null);
          }
          this.edges.add(row);
      }
      if (!edges.isEmpty) {
          edges.forEach((e) {
              edge(e[0], e[1], e[2]);
          });
      }
  }
  
  void edge(num from, num to, num weight) {
      edges[from - 1][to - 1] = weight;
  }
 
  Map _constructShortestPath(List<num> distances, List<num> previous, List<num> unvisited, num to) {
      num vertex = to;
      List<num> path = new List<num>();

      while (null != vertex) {
          path.add(vertex + 1);
          vertex = previous[vertex];
      };
      
      return {
         'path': path,
         'length': distances[to]
      };
  }
  
  num _getUnvisitedVertexWithShortestPath(List<num> distances, List<num> previous, List<num> unvisited) {
    num minimumDistance = 1/0;
    num vertex = null;
      
    unvisited.forEach((unvisitedVertex) {
      if (distances[unvisitedVertex] < minimumDistance) {
        vertex = unvisitedVertex;
        minimumDistance = distances[vertex];
      };
    });
    return vertex;
  }
  
  void _updateDistancesForCurrent(List<num> distances, List<num> previous, List<num> unvisited, num current) {  
    for (num i = 0; i < edges[current].length; i++) {
      num currentEdge = edges[current][i];
      
      if ((null != currentEdge) && unvisited.contains(i)) {
        if (distances[current] + currentEdge < distances[i]) {
          distances[i] = distances[current] + currentEdge;
          previous[i] = current;
        };
      };
    };
  }
  
  //Dijkstra algorithm http://en.wikipedia.org/wiki/Dijkstra's_algorithm
  Map getShortestPath(num from, num to) {  
      List<num> unvisited = new List<num>();
      num current = null;
      List<num> distances = new List<num>();
      List<num> previous = new List<num>();

      from = from - 1;    
      to = to - 1;
      //Initialization
      for (num i = 0; i < nodesNumber; i++) {
          unvisited.add(i);
          //Infinity
          distances.add(1/0);
          previous.add(null);
      };
      distances[from] = 0;
    
      while (true) {
          if (!unvisited.contains(to)) {
              return _constructShortestPath(distances, previous, unvisited, to);
          };
          current = _getUnvisitedVertexWithShortestPath(distances, previous, unvisited);
    
          //No path exists
          if ((null == current) || (1/0 == distances[current])) {
              return {
                  'path': [],
                  'length': 1/0
              };
          };
          this._updateDistancesForCurrent(distances, previous, unvisited, current);     
          unvisited.remove(current);
      };
  }
}

void main() {
    Graph graph = new Graph(8, [
        [1, 2, 5], [1, 3, 1], [1, 4, 3],
        [2, 3, 2], [2, 5, 2],
        [3, 4, 1], [3, 5, 8],
        [4, 6, 2],
        [5, 7, 1],
        [6, 5, 1]
    ]);
  
    Map shortestPath = graph.getShortestPath(1, 7);

    print("path = ");
    print(shortestPath['path'].join(","));
    print("length = ");
    print(shortestPath['length']);
  
    //No shortest path to the vertex '8'
    print(graph.getShortestPath(1, 8));
}